lib/std/secrets.nim

```diff
@@ -0,0 +1,226 @@
+  doAssert urandom(120).len == 120
+
+
+
+when defined(posix):
+  import std/posix
+
+  const batchSize = 256
+
+  template processReadBytes(readBytes: int, p: pointer, res: var int) =
+    if readBytes == 0:
+      break
+    elif readBytes > 0:
+      inc(res, readBytes)
+      cast[ptr pointer](p)[] = cast[pointer](cast[ByteAddress](p) + readBytes)
+    else:
+      if osLastError().int in {EINTR, EAGAIN}:
+        discard
+      else:
+        res = -1
+        break
+
+  proc getDevUrandom(p: var openArray[byte], size: Natural): int =
+    let size = p.len
+    if size == 0:
+      return
+
+    let fd = posix.open("/dev/urandom", O_RDONLY)
+
+    if fd > 0:
+      var stat: Stat
+      if fstat(fd, stat) != -1 and S_ISCHR(stat.st_mode):
+        let
+          chunks = (size - 1) div batchSize
+          left = size - chunks * batchSize
+
+        var base = 0
+        for i in 0 ..< chunks:
+          let readBytes = posix.read(fd, addr p[base], batchSize)
+          if readBytes < 0:
+            return readBytes
+          inc(base, batchSize)
+
+        result = posix.read(fd, addr p[base], left)
+      discard posix.close(fd)
+
+when defined(js):
+  import std/private/jsutils
+
+  proc getRandomValues(arr: Uint8Array) {.importjs: "window.crypto.getRandomValues(#)".}
+
+  const
+    maxBrowserCryptoBufferSize = 256
+
+  proc urandom*(p: var openArray[byte]): int =
+    let size = p.len
+    if size > 0:
+      let chunks = (size - 1) div maxBrowserCryptoBufferSize
+      var base = 0
+      for i in 0 ..< chunks:
+        for j in 0 ..< maxBrowserCryptoBufferSize:
+          var src = newUint8Array(maxBrowserCryptoBufferSize)
+          getRandomValues(src)
+          p[base + j] = src[j]
+
+        inc(base, maxBrowserCryptoBufferSize)
+
+      let left = size - chunks * maxBrowserCryptoBufferSize
+      var src = newUint8Array(left)
+      getRandomValues(src)
+      for i in 0 ..< left:
+        p[base + i] = src[i]
+
+elif defined(windows):
+  type
+    PVOID = pointer
+    BCRYPT_ALG_HANDLE = PVOID
+    PUCHAR = ptr cuchar
+    NTSTATUS = clong
+    ULONG = culong
+
+  const
+    STATUS_SUCCESS = 0x00000000
+    BCRYPT_USE_SYSTEM_PREFERRED_RNG = 0x00000002
+
+  proc bCryptGenRandom(
+    hAlgorithm: BCRYPT_ALG_HANDLE,
+    pbBuffer: PUCHAR,
+    cbBuffer: ULONG,
+    dwFlags: ULONG
+  ): NTSTATUS {.stdcall, importc: "BCryptGenRandom", dynlib: "Bcrypt.dll".}
+
+
+  proc randomBytes(pbBuffer: pointer, cbBuffer: int): int =
+    bCryptGenRandom(nil, cast[PUCHAR](pbBuffer), ULONG(cbBuffer),
+                            BCRYPT_USE_SYSTEM_PREFERRED_RNG)
+
+  proc urandom*(p: var openArray[byte]): int =
+    let size = p.len
+    if size > 0:
+      result = randomBytes(addr p[0], size)
+
+elif defined(linux):
+  let SYS_getrandom {.importc: "SYS_getrandom", header: "<syscall.h>".}: clong
+
+  proc syscall(n: clong, buf: pointer, bufLen: cint, flags: cuint): int {.importc: "syscall", header: "syscall.h".}
+
```

Comment:
why not https://man7.org/linux/man-pages/man2/getrandom.2.html ? 

> getrandom() was introduced in version 3.17 of the Linux kernel.
> Support was added to glibc in version 2.25.

Reply:
use `sys/syscall.h` is more safer. Because glibc 2.25 is a bit new

Reply:
Better solution is to check whether `getrandom` is available, I will do it later.

Reply:
See what Python does:
https://github.com/python/cpython/blob/15bd9efd01e44087664e78bf766865a6d2e06626/Python/bootstrap_hash.c#L13